Registry/RepositoryRegistry.php

```
@@ -0,0 +1,147 @@
+    }
+
+    /**
+     * Add a repository factory.
+     *
+     * @param string $name
+     * @param RepositoryFactoryInterface
+     */
+    public function addFactory($name, RepositoryFactoryInterface $factory)
+    {
+        if (isset($this->factories[$name])) {
+            throw new \RuntimeException(sprintf(
+                'A factory named "%s" has already been registered',
+                $name
+            ));
+        }
+
+        $this->factories[$name] = $factory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function get($instanceName)
+    {
+        if (!isset($this->instances[$instanceName])) {
+            $this->instances[$instanceName] = $this->createRepositoryInstance($instanceName);
+        }
+
+        return $this->instances[$instanceName];
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepositoryAlias(ResourceRepository $repository)
+    {
+        foreach ($this->instances as $alias => $repositoryInstance) {
+            if ($repositoryInstance === $repository) {
+                return $alias;
+            }
+        }
+
+        throw new \InvalidArgumentException(sprintf(
+            'Could not determine alias of repository of class "%s", it was not created by this registry.',
+            get_class($repository)
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getRepositoryType(ResourceRepository $repository)
+    {
+        $repositoryClass = get_class($repository);
+        if (!isset($this->typeMap[$repositoryClass])) {
+            throw new \InvalidArgumentException(sprintf(
+                'No repository has been instantiated of class "%s", cannot determine the type, known types: "%s".',
+                $repositoryClass,
+                implode('", "', $this->typeMap)
+            ));
+        }
+
+        return $this->typeMap[$repositoryClass];
+    }
+
+    private function createRepositoryInstance($instanceName)
```

why is this not included in the DI extension and then registered as a method call to `addFactory`? This allows us to fail fast when there is a misconfiguration and it allows the repository factories to use the TreeBuilder of the config component to explicitely define the options that can be configured.